research/context/preselection/preselect_wikipedia_comments.py

```diff
@@ -0,0 +1,176 @@
+    :return:
+    """
+    client = bigquery.Client(project=bq_project_name)
+    return client
+
+def query(low_tox, high_tox, limit, max_txt_len, min_txt_len, type, bq_table_name):
+    """
+    Google Big Query sampling data of specific types
+    :param low_tox: comments should be higher than that
+    :param high_tox: comments should be lower or equal than that
+    :param limit: fetch some
+    :param type: kind of comments (wikipedia specific)
+    :param max_txt_len: max length of comment
+    :param bq_table_name: the table name in BigQuery
+    :return: dataframe with results
+    """
+    assert low_tox>=0 and high_tox<=1
+    query = """
+    SELECT *, rand() as r 
+        FROM `{bq_table_name}`
+        WHERE length(cleaned_content)>{min_txt_len} AND length(cleaned_content)<{max_txt_len} AND replyTo_id!="null" AND type="{typ}" AND toxicity>={low} AND toxicity<{high}
+        ORDER BY r
+        LIMIT {lim}
+    """.format(bq_table_name=bq_table_name, low=low_tox, high=high_tox, lim=limit, typ=type, max_txt_len=max_txt_len, min_txt_len=min_txt_len)
+    query_job = client.query(query)
+    results = query_job.result()
+    return results.to_dataframe()
+
+
+def sample_based_on_toxicity(limit, max_txt_len, min_txt_len, type, bq_table_name, toxic_zones_num, randomly_selected_limit):
+    """
+    Google Big Query sampling data of specific types
+    :param limit: fetch some
+    :param type: kind of comments (wikipedia specific)
+    :param max_txt_len: max length of comment
+    :param bq_table_name: the table name in BigQuery
+    :param toxic_zones_num: number of toxic zones to be created from 0 to 1
+    :return: dataframe with results
+    """
+    query = """"""
+    if randomly_selected_limit>0:
+        query = """
+            (SELECT *, rand() as r 
+            FROM `{bq_table_name}`
+            WHERE length(cleaned_content)>{min_txt_len} AND length(cleaned_content)<{max_txt_len} AND replyTo_id!="null" AND type="{typ}" 
+            LIMIT {limit} 
+            )
+            """.format(bq_table_name=bq_table_name, typ=type, max_txt_len=max_txt_len, min_txt_len=min_txt_len, limit=randomly_selected_limit)
+    for i in range(toxic_zones_num):
+        low_tox_zone = float(i) / toxic_zones_num
+        high_tox_zone = float(i + 1) / toxic_zones_num
+        new_query = """
+            (SELECT *, rand() as r 
+            FROM `{bq_table_name}`
+            WHERE length(cleaned_content)>{min_txt_len} AND length(cleaned_content)<{max_txt_len} AND replyTo_id!="null" AND type="{typ}" AND toxicity>={low} AND toxicity<{high}
+            LIMIT {limit})
+            """.format(bq_table_name=bq_table_name, low=low_tox_zone, high=high_tox_zone, typ=type, max_txt_len=max_txt_len, min_txt_len=min_txt_len, limit=limit)
+        if len(query)==0: # if no random query done already
+            query = new_query
+        else:
+            query += """\n\nUNION ALL\n\n""" + new_query # else, unite the queries
```

Comment:
Instead of doing a incremental string concat, you'll get cleaner code if you add to a list each query, and then use a join to insert the "UNION ALL" parts. This is part of a general pattern of try to create and work with the best data type representation of your real data. List of queries is the right datatype; and only at the end throw away information / push information into a different uniform object type.